### ISSUE TITLE
Petition Gradle to support JVM-local configuration-cache

### ISSUE THREAD

**Issue by nedtwigg** [diffplug/spotless#987]
**TLDR: Unlike remote build cache, configuration cache is never going to be relocatable. Because of that, it might as well be a JVM-local cache (within-daemon), which removes the performance and incidental-complexity penalties of a serialization roundtrip. Spotless #980 hacks a workaround to support this. But anytime you start a new daemon, you need to `rm -rf .gradle/configuration-cache` to rehydrate the non-serializable parts of Spotless. Gradle could support this natively so that configuration cache is faster and easier to adopt.**

One of Gradle's best features is the [build cache](https://docs.gradle.org/current/userguide/build_cache.html) (`--build-cache` or `org.gradle.caching=true`). It snapshots the input properties of a task and memoizes the outputs. Simple and effective. Best of all, cache artifacts generated on a single CI server can be distributed to all other builds thanks to "remote build cache". However, in order for that relocation to work, you need to annotate every file-based input with one of four [`PathSensitivity`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/PathSensitivity.html) values: NONE, NAME_ONLY, RELATIVE, ABSOLUTE. There's really a fifth value too, which is "unannotated, does not support relocating (neither to other machines nor to other folders on this machine)."

In 6.6, Gradle shipped [configuration cache](https://docs.gradle.org/current/userguide/configuration_cache.html) (`--configuration-cache` or `org.gradle.unsafe.configuration-cache=true`). It places [lots of restrictions](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements) on how tasks can operate and coordinate. For Spotless, complying with the simple part of these restrictions required ~150 lines of churn plus another ~200 lines of boilerplate (#982).

But the really tricky restriction, on both correctness and performance, is that all tasks must be able to roundtrip their state losslessly through Java's built-in semi-deprecated serialization system. For Spotless that would require something close to a rewrite, but we were able to hack around it with ~300 lines of trickery (#986).

## The advantage of serialization

- It is "proof-of-isolation". If you can put yourself onto disk and take yourself back out, then you have proven that you are isolated from other tasks, which makes it safe to run tasks in parallel by default.

## The disadvantages of serialization

- Slow.
- Many useful things cannot be serialized.
- It isn't "real" serialization, there are [several caveats](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization) to what parts of Java serialization Gradle will let you use.

## Sharing configuration cache between machines is going to be error-prone and labor-intensive

There is a feature proposal (https://github.com/gradle/gradle/issues/13510) to share configuration cache information between machines, same idea as remote build cache. Remember that to make that work for the remote build cache, you needed to annotate every input property with a `PathSensitivity` value. The equivalent for remote configuration cache is for every POJO which contains a system path to be annotated. Given that Gradle isn't using "real" serialization, I suppose they may have a workable plan for this, where you can somehow scope that "any File reference below this is PathSensitivty.NAME_ONLY" or something like that. But it's even harder than that, because build cache didn't need to resynthesize its inputs, it just needed to one-way fingerprint them. What does it mean to take a `File implementationDetailFile` through a roundtrip of `PathSensitivity.NAME_ONLY`? This is especially problematic because the *easy* part of complying with configuration cache is to convert every call to `project.projectDir` into:

> A task input or output property or a script variable to capture the result of using project.projectDir to calculate the actual parameter.

How is a future version of Gradle going to know which need to be rewritten machine to machine, versus which are intended to be left alone? Plus, who cares! Build cache is already effective at offloading work from developer machines to CI machines. The real gold of configuration cache is improving iteration times on a single dev machine (time-to-test), and for that environment you'd happily trade away relocatability to get the fastest possible iteration time. The fastest possible iteration time is not going to have a serialization roundtrip as one of the required steps.
 
## Sharing configuration cache between *builds* on a single machine is faster and easier

Take an API like the following:

```java
// user-facing
org.gradle.api.CacheKey<T> { T get(); void set(T); }

class MyServerConnection {} // implements .equals and .hashCode, but not serializable
class MyTask {
  private final CacheKey<ServerConnection> serverConnection = createCacheKey<>("serverConnection")

  public void setConnection(ServerConnection connection) {
    serverConnection.set(connection);
  }

  @Input
  public ServerConnection getConnection() {
    return serverConnection.get();
  }
  // field/getter/setter could be replaced with @Inject CacheableProperty<ServerConnection> getConnection()

  @TaskAction
  public void performAction(InputChanges inputs) throws Exception {
    serverConnection.get().doStuff()
  }
}

// implementation
data class InternalCacheKey {
  File projectDir
  String taskPath
  String stateName
}

private static Map<InternalCacheKey, Object> daemonGlobalState = ...

class GradleInternalTaskImplementation implements Task {
  CacheKey<T> createCacheKey(String name) {
    InternalCacheKey internalKey = new InternalCacheKey(projectDir, getPath(), name);
    return new CacheKey<T>() {
      @Override public void set(T value) {
         daemonGlobalState.put(internalKey, value);
       }
      @Override public T get() {
         Object value = daemonGlobalState.get(internalKey);
          if (value == null) {
            throw TriggerConfigurationException(); // we need to rerun configuration to populate our cache
          } else {
            return (T) value;
          }
       }
    }
  }
```

This is how we implemented configuration cache for Spotless, except that we have no way to do `TriggerConfigurationException`. See #986.

**Comment by nedtwigg** [comment]
One of Gradle's best features is the [build cache](https://docs.gradle.org/current/userguide/build_cache.html) (`--build-cache` or `org.gradle.caching=true`). It snapshots the input properties of a task and memoizes the outputs. Simple and effective. Best of all, cache artifacts generated on a single CI server can be distributed to all other builds thanks to "remote build cache". However, in order for that relocation to work, you need to annotate every file-based input with one of four [`PathSensitivity`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/PathSensitivity.html) values: NONE, NAME_ONLY, RELATIVE, ABSOLUTE. There's really a fifth value too, which is "unannotated, does not support relocating (neither to other machines nor to other folders on this machine)."

In 6.6, Gradle shipped [configuration cache](https://docs.gradle.org/current/userguide/configuration_cache.html) (`--configuration-cache` or `org.gradle.unsafe.configuration-cache=true`). It places [lots of restrictions](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements) on how tasks can operate and coordinate. For Spotless, complying with the simple part of these restrictions required ~150 lines of churn plus another ~200 lines of boilerplate (#982).

But the really tricky restriction, on both incidental complexity and performance, is that all tasks must be able to roundtrip their state losslessly through Java's semi-deprecated serialization system. For Spotless that would require something close to a rewrite, but we were able to hack around it with ~300 lines of trickery (#986). That trickery is much faster and easier to adopt than Gradle's current approach, so we hope that they will switch.

## The advantage of serialization

- It is a "proof-of-isolation" (but not the only one!). If you can put yourself onto disk and take yourself back out, then you have proven that you are isolated from other tasks, which makes it safe to run tasks in parallel by default.

## The disadvantages of serialization

- Slow.
- Many useful things cannot be serialized.
- It isn't "real" serialization, there are [several caveats](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization) to what parts of Java serialization Gradle will let you use.

## Sharing configuration cache between machines is going to be error-prone and labor-intensive

There is a feature proposal (https://github.com/gradle/gradle/issues/13510) to share configuration cache information between machines, same idea as remote build cache. Remember that to make that work for the remote build cache, you needed to annotate every input property with a `PathSensitivity` value. The equivalent for remote configuration cache is for every POJO which contains a system path to be annotated. Given that Gradle isn't using "real" serialization, I suppose they may have a workable plan for this, where you can somehow scope that "any File reference below this is PathSensitivty.NAME_ONLY" or something like that. But it's even harder than that, because build cache didn't need to resynthesize its inputs, it just needed to one-way fingerprint them. What does it mean to take a `File implementationDetailFile` through a roundtrip of `PathSensitivity.NAME_ONLY`? This is especially problematic because the *easy* part of complying with configuration cache is to convert every call to `project.projectDir` into:

> A task input or output property or a script variable to capture the result of using project.projectDir to calculate the actual parameter.

How is a future version of Gradle going to know which need to be rewritten machine to machine, versus which are intended to be left alone? Plus, who cares! Build cache is already effective at offloading work from developer machines to CI machines. The real gold of configuration cache is improving iteration times on a single dev machine (time-to-test), and for that environment you'd happily trade away relocatability to get the fastest possible iteration time. The fastest possible iteration time is not going to have a serialization roundtrip as one of the required steps.

**Comment by nedtwigg** [comment]
## Sharing configuration cache between *builds* on a single machine is faster and easier

Take an API like the following:

```java
// user-facing
org.gradle.api.CacheKey<T> { T get(); void set(T); }

class MyServerConnection {} // implements .equals and .hashCode, but not serializable
class MyTask {
  private final CacheKey<ServerConnection> serverConnection = createCacheKey<>("serverConnection")

  public void setConnection(ServerConnection connection) {
    serverConnection.set(connection);
  }

  @Input
  public ServerConnection getConnection() {
    return serverConnection.get();
  }
  // field/getter/setter could be replaced with @Inject CacheableProperty<ServerConnection> getConnection()

  @TaskAction
  public void performAction(InputChanges inputs) throws Exception {
    serverConnection.get().doStuff()
  }
}

// implementation
data class InternalCacheKey {
  File projectDir
  String taskPath
  String stateName
}

private static Map<InternalCacheKey, Object> daemonGlobalState = ...

class GradleInternalTaskImplementation implements Task {
  CacheKey<T> createCacheKey(String name) {
    InternalCacheKey internalKey = new InternalCacheKey(projectDir, getPath(), name);
    return new CacheKey<T>() {
      @Override public void set(T value) {
         daemonGlobalState.put(internalKey, value);
       }
      @Override public T get() {
         Object value = daemonGlobalState.get(internalKey);
          if (value == null) {
            throw TriggerConfigurationException(); // we need to rerun configuration to populate our cache
          } else {
            return (T) value;
          }
       }
    }
  }
```

This is how we implemented configuration cache for Spotless, except that we have no way to do `TriggerConfigurationException`. See #986.

**Comment by eskatos** [comment]
Hi Ned,

Thanks for bringing this up.

First, with regards to the configuration-cache serialization: it is not based on Java Serialization - we are not that crazy :) Gradle uses its own serialization mechanism and format for configuration cache entries, it is optimized for i/o, computation and space saving. There is some support for Java Serialization as a fallback. But it is not recommended relying on it and handcrafting serialization. I just tried to make this clearer in the docs in https://github.com/gradle/gradle/pull/19005

We may implement an in-memory cache at some point, for example we’ve talked about pre-creating the task graph at the end of the build in anticipation of the next build. But we think it would still make copies of any mutable structures (for isolation) and would still be backed by a persistent copy (for daemon restarts). Isolating tasks enables running all of them in parallel safely. Using shared mutable state wouldn't work.  We’d need some data that shows that the current approach actually has enough of a negative impact on build performance to do something about it (not saying it doesn’t, only that we don’t yet know whether it does or does not).

In terms of relocatabilty, let’s see. We’ve talked about using the remote cache, but we have some work to do before that would be possible. It’s hard to see us doing this without adding some more constraints or asking you to be more explicit about relocation.

**Comment by nedtwigg** [comment]
If the cache is only shared between sequential builds, then it is not *shared* mutable state. Just mutable state. And that's a risky thing, but not so risky if you've opted-in to it.

And if Gradle never builds that API, that's okay, because it's very easy to do in userspace.

The only part we can't build in userspace is `TriggerConfigurationError`. Users will have to do that manually until/unless Gradle builds a way for tasks to say "my cache is cold, I need this build to be a slow reconfigure". I suppose we could workaround that in userspace with something like `./gradlew2` that checks output for `TriggerConfigurationError` and then does `rm -rf .gradle/configurtation-cache` and reruns the originally requested task...

I think our biggest disagreement is around user story. IMO, the dominant user story is a serial chain of `edit -> test -> edit -> test`. That story has some built-in constraints, and I would take advantage of those constraints by designing to them. Happy to agree to disagree because we've already got the primitives we need :)

**Comment by eskatos** [comment]
I meant state shared between tasks.

**Comment by nedtwigg** [comment]
I might be naive, but that seems like another "design to the wrong point" problem. The worrycase as I understand it is a poorly implemented userland plugin which:

1. defines multiple tasks...
2. which depend on each other enough to share mutable state...
3. but also don't `dependOn` each other, such that they are eligible to run concurrently...
4. and this mutable state was improperly implemented

I agree it's possible, but it doesn't seem like it's so common that it's worth designing to. Especially when the builds where correctness matters the most are CI builds where configuration-cache doesn't make much sense in the first place.

Let's say that Gradle came with a built-in `./gradlew taskFuzz` which

1. performs a no-parallel build and checkpoints the output artifacts
2. looks at the dag to find which tasks could possibly run concurrently with each other
3. bashes on each pair of conceivably-concurrent tasks to see if it can get a bug or at least modify the output
4. outputs a `.gradle-parallel-safe` file which is a hash of the buildscripts that certifies that this build is safe to run concurrently

With some kind of verification tool, the implementation is no longer responsible for verification, and you open up a lot of design space for the implementation.

The current "verification" tool (roundtrip serialization baked into every execution) is trivially defeated with a static variable containing mutable state. In Spotless, we are doing this on purpose, but it's easy to do this on accident too. A serialization roundtrip won't be able to catch bugs of this type. You could add some classloader magic, but now plugin authors are going to be debugging not only magic serialization, but also magic classloading. You'll also be paying another performance penalty by thrashing the JIT.

With each release of Gradle, I would hope that authoring builds will get easier, and that I will spend less time on the incidental complexity of the jobs I want Gradle to do. My concern with the configuration-cache is that there is a big set of stuff that is fundamentally not serializable, and any build with those things in it makes the entire build a second-class experience. Besides Spotless, I have updated four other public build plugins for configuration-cache, and more private ones:

- https://github.com/diffplug/image-grinder/pull/9
- https://github.com/diffplug/spotless-changelog/pull/27
- https://github.com/diffplug/blowdryer/pull/26
- https://github.com/diffplug/goomph/pull/153

In every case, I am forced to write quite a lot of boilerplate, and formerly straightforward things become quite complicated (e.g. reading project / environment variables).

I would love a feature like "developer-cache", marketed as so:

- Define your build logic in the simplest and most straightforward way that you can, and we'll make sure that the `edit -> test -> edit -> test` cycle time is as fast as possible.

A "developer-cache" design would be unwilling to incur any runtime costs on verification responsibilities, and it wouldn't spend developer time on relocatability.

**Comment by JavierSegoviaCordoba** [comment]
Have you thought to implement the "automatic mode" for deleting the configuration cache?

When I tried this feature it was failing mainly in CI which is annoying.

Is it not possible to change the implementation to be really compatible with the Gradle way?

**Comment by nedtwigg** [comment]
> Have you thought to implement the "automatic mode" for deleting the configuration cache?

- Good idea: #1209

> it was failing mainly in CI which is annoying.

Have you looked at what kind of speedup/slowdown you get if you disable configuration cache on the CI machines? If the inbound CI queue to a single runner is even just a little bit heterogenous, you can easily spend more time writing `configuration-cache` entries than you save from reading them back later. As opposed to build-cache, which is a fantastic no-downside speedup for CI machines. I believe Gradle recommends to disable configuration cache for CI machines ([docs](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:adoption)). 

> Is it not possible to change the implementation to be really compatible with the Gradle way

We tried this in #644, and it's pretty close to a rewrite for us, [for this reason](https://github.com/diffplug/spotless/issues/644#issuecomment-714682531). I'd be okay to merge PRs in that direction, but it would be a lot of work, and at the end of all the work you'd be a lot slower than a within-daemon JVM-local cache.

**Comment by JavierSegoviaCordoba** [comment]
> Have you looked at what kind of speedup/slowdown you get if you disable configuration cache on the CI machines? 

I haven't tried it personally, but looks like having it enabled is the direction that tools like Gradle Build Action are taking.

**Comment by BraisGabin** [comment]
We were seeing a lot of this message:

> Spotless JVM-local cache is stale. Regenerate the cache with
>   `rm -rf .gradle/configuration-cache`

So we end up disabling the configuration cache for `SpotlessTask`. We know that this will slow down our executions but we prefer that than get "random" errors.

PLEASE DON'T COPY&PASTE THIS IF YOU DON'T KNOW WHAT IT DOES!

```kotlin
tasks.withType<SpotlessTask>().configureEach {
    notCompatibleWithConfigurationCache("https://github.com/diffplug/spotless/issues/987")
}
```

**Comment by melix** [comment]
I am just trying to make the Micronaut build compatible with the configuration cache and Spotless compatibility is the last missing piece. I had to disable the cache with the technique @BraisGabin demonstrated, but this is far from ideal, since basically any `check` build will not benefit from the cache and display a _huge_ message at the end of the build: 

```
Configuration cache entry discarded because incompatible tasks were found: ':micronaut-test-resources-hibernate-reactive-oracle-xe:spotlessJava', ':micronaut-test-resources-r2dbc-mysql:spotlessJavaMisc', ':micronaut-test-resources-testcontainers:spotlessJavaMisc', ':micronaut-test-resources-r2dbc-postgresql:spotlessJavaMisc', ':micronaut-test-resources-core:spotlessJava', ':micronaut-test-resources-jdbc-postgresql:spotlessJava', ':micronaut-test-resources-r2dbc-mssql:spotlessJavaMisc', ':micronaut-test-resources-hibernate-reactive-core:spotlessJavaMisc', ':micronaut-test-resources-jdbc-oracle-xe:spotlessJava', ':micronaut-test-resources-core:spotlessJavaMisc', ':micronaut-test-resources-jdbc-core:spotlessJavaMisc', ':micronaut-test-resources-r2dbc-mysql:spotlessJava', ':micronaut-test-resources-r2dbc-oracle-xe:spotlessJavaMisc', ':micronaut-test-resources-hibernate-reactive-oracle-xe:spotlessJavaMisc', ':micronaut-test-resources-build-tools:spotlessJava', ':micronaut-test-resources-neo4j:spotlessJava', ':micronaut-test-resources-jdbc-postgresql:spotlessJavaMisc', ':micronaut-test-resources-localstack-core:spotlessJava', ':micronaut-test-resources-jdbc-oracle-xe:spotlessJavaMisc', ':micronaut-test-resources-jdbc-mysql:spotlessJavaMisc', ':micronaut-test-resources-hibernate-reactive-mssql:spotlessJavaMisc', ':micronaut-test-resources-hivemq:spotlessJava', ':micronaut-test-resources-rabbitmq:spotlessJavaMisc', ':micronaut-test-resources-localstack-dynamodb:spotlessJavaMisc', ':micronaut-test-resources-r2dbc-pool:spotlessJava', ':micronaut-test-resources-hibernate-reactive-core:spotlessJava', ':micronaut-test-resources-hibernate-reactive-mariadb:spotlessJavaMisc', ':micronaut-test-resources-embedded:spotlessJavaMisc', ':micronaut-test-resources-localstack-s3:spotlessJavaMisc', ':micronaut-test-resources-hivemq:spotlessJavaMisc', ':micronaut-test-resources-redis:spotlessJavaMisc', ':micronaut-test-resources-r2dbc-core:spotlessJavaMisc', ':micronaut-test-resources-r2dbc-core:spotlessJava', ':micronaut-test-resources-localstack-s3:spotlessJava', ':micronaut-test-resources-mongodb:spotlessJava', ':micronaut-test-resources-testcontainers:spotlessJava', ':micronaut-test-resources-localstack-core:spotlessJavaMisc', ':micronaut-test-resources-jdbc-mysql:spotlessJava', ':micronaut-test-resources-r2dbc-oracle-xe:spotlessJava', ':micronaut-test-resources-server:spotlessJavaMisc', ':micronaut-test-resources-hibernate-reactive-mysql:spotlessJavaMisc', ':micronaut-test-resources-jdbc-mssql:spotlessJava', ':micronaut-test-resources-r2dbc-mariadb:spotlessJava', ':micronaut-test-resources-localstack-dynamodb:spotlessJava', ':micronaut-test-resources-client:spotlessJavaMisc', ':micronaut-test-resources-hibernate-reactive-postgresql:spotlessJavaMisc', ':micronaut-test-resources-kafka:spotlessJava', ':micronaut-test-resources-embedded:spotlessJava', ':micronaut-test-resources-hashicorp-vault:spotlessJava', ':micronaut-test-resources-mongodb:spotlessJavaMisc', ':micronaut-test-resources-r2dbc-postgresql:spotlessJava', ':micronaut-test-resources-r2dbc-pool:spotlessJavaMisc', ':micronaut-test-resources-client:spotlessJava', ':micronaut-test-resources-r2dbc-mariadb:spotlessJavaMisc', ':micronaut-test-resources-elasticsearch:spotlessJava', ':micronaut-test-resources-r2dbc-mssql:spotlessJava', ':micronaut-test-resources-neo4j:spotlessJavaMisc', ':micronaut-test-resources-build-tools:spotlessJavaMisc', ':micronaut-test-resources-jdbc-mssql:spotlessJavaMisc', ':micronaut-test-resources-kafka:spotlessJavaMisc', ':micronaut-test-resources-rabbitmq:spotlessJava', ':micronaut-test-resources-jdbc-mariadb:spotlessJavaMisc', ':micronaut-test-resources-elasticsearch:spotlessJavaMisc', ':micronaut-test-resources-jdbc-mariadb:spotlessJava', ':micronaut-test-resources-hibernate-reactive-mariadb:spotlessJava', ':micronaut-test-resources-hibernate-reactive-mysql:spotlessJava', ':micronaut-test-resources-hibernate-reactive-mssql:spotlessJava', ':micronaut-test-resources-redis:spotlessJava', ':micronaut-test-resources-jdbc-core:spotlessJava', ':micronaut-test-resources-server:spotlessJava', ':micronaut-test-resources-hashicorp-vault:spotlessJavaMisc', ':micronaut-test-resources-hibernate-reactive-postgresql:spotlessJava'.
```

So I would personally really appreciated if you adapted the plugin.

**Comment by nedtwigg** [comment]
PR's welcome :)

- #1274.

**Comment by pbielicki** [comment]
To me this issue has no sense. The configuration cache is meant to be reused locally by any daemon. Gradle daemon can disappear at any time and having a JVM-only cached configuration is just rubbish (personal opinion not meant to offend anyone).

IMO there is a design flaw in the spotless plugin. I tried to make it CC compatible but the code needs to be seriously re-architectured. Currently, spotless creates a lot of JVM-lazy (not Gradle-lazy) objects e.g. lambdas that define behaviours of concrete formatters. These lambdas are not serializable in any way and that's the problem. 
In order for the CC to work Gradle needs to be able to store and restore the configuration in the subsequent executions. It's not possible with the current spotless design.

The way to move forward, as I see it, is to delay creation of those lambdas until the Gradle execution phase. Spotless configuration should only know which formatters to use (e.g. scala or kotlin) and where the formatter configuration file is located.

It's not an easy-peasy change, as it requires quite some changes in the spotless codebase.

**Comment by nedtwigg** [comment]
I'm super bummed that serialization roundtrip is a base requirement for Gradle plugins. IMO it's a drag on plugin development and maintenance, as well as execution time. I really wish Gradle plugins had access to a JVM-local cache designed for serial development. Try something, run a test, try something, run a test; that's the loop that I want to optimize for.

But Gradle is clearly heavily committed to this path at this point, and we finally have a reasonable path forward for supporting roundtrip serialization. If you'd like to help:

- wait for #1945 to get merged
- grab a step from https://github.com/diffplug/spotless/issues/1274#issuecomment-1839694669 and make it roundtrip serializable, open a draft PR and link it in the issue when you have any progress

**Comment by xenoterracide** [comment]
why is it always `lineEndingsPolicy`, at least it seems to be for me. That would come from `.editorconfig` which hasn't changed in quite some time.

```
> Error while evaluating property 'lineEndingsPolicy' of task ':spotlessKotlinGradle'.
   > Spotless JVM-local cache is stale. Regenerate the cache with
       rm -rf .gradle/configuration-cache
```
